Raise Ruby exception in JRuby

I would like to raise an exception class defined in Ruby from Java code. How would one do this?

The examples given in https://github.com/jruby/jruby/wiki/JRuby-Reference#Exceptions all show how to work with Java exceptions, but I would like to raise a Ruby exception.

We have this code:

<pre>
    try {
      string = (RubyString) value;
    } catch (ClassCastException e) {
      throw context.runtime.newArgumentError(e.toString());
    }
</pre>

I tried:

<pre>
        RubyClass cls = this.runTime.getClass("Mongo::GssapiNative::Error");
        RubyString msg = RubyString.newString(context.runtime, message);
        throw cls.newInstance(context, msg, Block.NULL_BLOCK);
</pre>

This failed with:

<pre>
src/main/org/mongodb/sasl/GSSAPIAuthenticator.java:141: error: incompatible types: IRubyObject cannot be converted to Throwable
        throw cls.newInstance(context, msg, Block.NULL_BLOCK);
                             ^
</pre>

So I need to convert my Ruby class instance to a Throwable?

Related question: our [existing code](https://github.com/mongodb/mongo-ruby-kerberos/blob/master/src/main/org/mongodb/sasl/MongoSecurityException.java) supports exception causes, how do I attach those to Ruby exception classes? Is this handled for me automatically somehow?

Thanks in advance.

--- a/core/src/main/java/org/jruby/Ruby.java
+++ b/core/src/main/java/org/jruby/Ruby.java
@@ -4243,9 +4243,14 @@ public final class Ruby implements Constantizable {
     }
 
     /**
-     * @param exceptionClass
-     * @param message
-     * @return
+     * Construct a new RaiseException wrapping a new Ruby exception object appropriate to the given exception class.
+     *
+     * There are additional forms of this construction logic in {@link RaiseException#from}.
+     *
+     * @param exceptionClass the exception class from which to construct the exception object
+     * @param message a simple message for the exception
+     * @return a new RaiseException wrapping a new Ruby exception
+     * @see RaiseException#from(Ruby, RubyClass, String)
      */
     public RaiseException newRaiseException(RubyClass exceptionClass, String message) {
         return RaiseException.from(this, exceptionClass, message);

--- a/core/src/main/java/org/jruby/RubyException.java
+++ b/core/src/main/java/org/jruby/RubyException.java
@@ -369,6 +369,11 @@ public class RubyException extends RubyObject {
     public void setCause(IRubyObject cause) {
         this.cause = cause;
 
+        // don't do anything to throwable for null/nil cause to avoid forcing backtrace
+        // * if we have no cause yet, it's a no-op
+        // * if we have a cause, we can't change it
+        if (cause == null || cause.isNil()) return;
+
         Throwable t = toThrowable();
         Object javaCause;
 

--- a/core/src/main/java/org/jruby/RubyException.java
+++ b/core/src/main/java/org/jruby/RubyException.java
@@ -201,13 +201,28 @@ public class RubyException extends RubyObject {
         return (RubyException) exceptionClass.newInstance(runtime.getCurrentContext(), RubyString.newString(runtime, msg), Block.NULL_BLOCK);
     }
 
+    /**
+     * Construct a new RubyException object from the given exception class and message.
+     *
+     * @param context the current thread context
+     * @param exceptionClass the exception class
+     * @param message the message for the exception
+     * @return the new exception object
+     */
     public static RubyException newException(ThreadContext context, RubyClass exceptionClass, RubyString message) {
         return (RubyException) exceptionClass.newInstance(context, message, Block.NULL_BLOCK);
     }
 
-    @Deprecated
-    public static IRubyObject newException(ThreadContext context, RubyClass exceptionClass, IRubyObject message) {
-        return newException(context, exceptionClass, message.convertToString());
+    /**
+     * Construct a new RubyException object from the given exception class and message.
+     *
+     * @param context the current thread context
+     * @param exceptionClass the exception class
+     * @param args the arguments for the exception's constructor
+     * @return the new exception object
+     */
+    public static RubyException newException(ThreadContext context, RubyClass exceptionClass, IRubyObject... args) {
+        return (RubyException) exceptionClass.newInstance(context, args, Block.NULL_BLOCK);
     }
 
     @JRubyMethod
@@ -488,5 +503,10 @@ public class RubyException extends RubyObject {
         if (backtrace.backtraceData == null) {
             backtrace.backtraceData = context.runtime.getInstanceConfig().getTraceType().getIntegratedBacktrace(context, javaTrace);
         }
+    }
+
+    @Deprecated
+    public static IRubyObject newException(ThreadContext context, RubyClass exceptionClass, IRubyObject message) {
+        return newException(context, exceptionClass, message.convertToString());
     }
 }

--- a/core/src/main/java/org/jruby/RubyException.java
+++ b/core/src/main/java/org/jruby/RubyException.java
@@ -353,6 +353,17 @@ public class RubyException extends RubyObject {
 
     public void setCause(IRubyObject cause) {
         this.cause = cause;
+
+        Throwable t = toThrowable();
+        Object javaCause;
+
+        if (t.getCause() == null) {
+            if (cause instanceof RubyException) {
+                t.initCause(((RubyException) cause).toThrowable());
+            } else if (cause instanceof ConcreteJavaProxy && (javaCause = ((ConcreteJavaProxy) cause).getObject()) instanceof Throwable) {
+                t.initCause((Throwable) javaCause);
+            }
+        }
     }
 
     // NOTE: can not have IRubyObject as NativeException has getCause() returning Throwable

--- a/core/src/main/java/org/jruby/exceptions/Exception.java
+++ b/core/src/main/java/org/jruby/exceptions/Exception.java
@@ -27,6 +27,7 @@
 package org.jruby.exceptions;
 
 import org.jruby.RubyException;
+import org.jruby.javasupport.JavaUtil;
 
 /**
  * Represents a Ruby Exception as a throwable Java exception.
@@ -36,5 +37,19 @@ import org.jruby.RubyException;
 public class Exception extends RaiseException {
     public Exception(String message, RubyException exception) {
         super(message, exception);
+    }
+
+    @Override
+    public synchronized Throwable initCause(Throwable cause) {
+        Throwable t = super.initCause(cause);
+
+        // if init was successful, set Ruby exception's cause as well
+        if (cause instanceof Exception) {
+            getException().setCause(((Exception) cause).getException());
+        } else {
+            getException().setCause(JavaUtil.convertJavaToUsableRubyObject(getException().getRuntime(), cause));
+        }
+
+        return t;
     }
 }

--- a/core/src/main/java/org/jruby/exceptions/RaiseException.java
+++ b/core/src/main/java/org/jruby/exceptions/RaiseException.java
@@ -72,19 +72,94 @@ public class RaiseException extends JumpException {
         return this; // do not auto-fill from Throwable.<init> we fill (setStackTrace) in preRaise(...)
     }
 
-    @Deprecated
-    public static RaiseException from(RubyException exception, IRubyObject backtrace) {
-        return new RaiseException(exception, backtrace);
-    }
-
+    /**
+     * Construct a new throwable RaiseException appropriate for the target Ruby exception class.
+     *
+     * @param runtime the current JRuby runtime
+     * @param exceptionClass the class of the exception to construct and raise
+     * @param msg a simple message for the exception
+     * @return a RaiseException instance appropriate for the target Ruby exception class
+     */
     public static RaiseException from(Ruby runtime, RubyClass exceptionClass, String msg) {
         return RubyException.newException(runtime, exceptionClass, msg).toThrowable();
     }
 
+    /**
+     * Construct a new throwable RaiseException appropriate for the target Ruby exception class.
+     *
+     * @param runtime the current JRuby runtime
+     * @param exceptionClass the class of the exception to construct and raise
+     * @param msg a simple message for the exception
+     * @param backtrace a Ruby object (usually an Array) to use for the exception's backtrace
+     * @return a RaiseException instance appropriate for the target Ruby exception class
+     */
     public static RaiseException from(Ruby runtime, RubyClass exceptionClass, String msg, IRubyObject backtrace) {
         RubyException exception = RubyException.newException(runtime, exceptionClass, msg);
         exception.setBacktrace(backtrace);
         return exception.toThrowable();
+    }
+
+    /**
+     * Construct a new throwable RaiseException appropriate for the target Ruby exception class.
+     *
+     * @param runtime the current JRuby runtime
+     * @param exceptionClass the class of the exception to construct and raise
+     * @param args the arguments for the exception's constructor
+     * @return a RaiseException instance appropriate for the target Ruby exception class
+     */
+    public static RaiseException from(Ruby runtime, RubyClass exceptionClass, IRubyObject... args) {
+        RubyException exception = RubyException.newException(runtime.getCurrentContext(), exceptionClass, args);
+        return exception.toThrowable();
+    }
+
+    /**
+     * Construct a new throwable RaiseException appropriate for the target Ruby exception class.
+     *
+     * @param runtime the current JRuby runtime
+     * @param exceptionPath a string representing the fully-qualified constant path to look up the exception
+     * @param msg a simple message for the exception
+     * @return a RaiseException instance appropriate for the target Ruby exception class
+     */
+    public static RaiseException from(Ruby runtime, String exceptionPath, String msg) {
+        RubyClass exceptionClass = findExceptionClass(runtime, exceptionPath);
+        return from(runtime, exceptionClass, msg);
+    }
+
+    /**
+     * Construct a new throwable RaiseException appropriate for the target Ruby exception class.
+     *
+     * @param runtime the current JRuby runtime
+     * @param exceptionPath a string representing the fully-qualified constant path to look up the exception
+     * @param msg a simple message for the exception
+     * @param backtrace a Ruby object (usually an Array) to use for the exception's backtrace
+     * @return a RaiseException instance appropriate for the target Ruby exception class
+     */
+    public static RaiseException from(Ruby runtime, String exceptionPath, String msg, IRubyObject backtrace) {
+        RubyClass exceptionClass = findExceptionClass(runtime, exceptionPath);
+        return from(runtime, exceptionClass, msg, backtrace);
+    }
+
+    /**
+     * Construct a new throwable RaiseException appropriate for the target Ruby exception class.
+     *
+     * @param runtime the current JRuby runtimeString exceptionPath
+     * @param args the arguments for the exception's constructor
+     * @return a RaiseException instance appropriate for the target Ruby exception class
+     */
+    public static RaiseException from(Ruby runtime, String exceptionPath, IRubyObject... args) {
+        RubyClass exceptionClass = findExceptionClass(runtime, exceptionPath);
+        return from(runtime, exceptionClass, args);
+    }
+
+    private static RubyClass findExceptionClass(Ruby runtime, String exceptionPath) {
+        IRubyObject exceptionClass = runtime.getObject().getConstant(exceptionPath);
+
+        if (exceptionClass == null) {
+            throw runtime.newNameError("exception class not found", exceptionPath);
+        } else if (!(exceptionClass instanceof RubyClass)) {
+            throw runtime.newTypeError("expected to find exception class for " + exceptionPath + " but got " + exceptionClass.inspect());
+        }
+        return (RubyClass) exceptionClass;
     }
 
     @Override
@@ -269,6 +344,11 @@ public class RaiseException extends JumpException {
         } else {
             setStackTrace(skipFillInStackTracePart(javaTrace));
         }
+    }
+
+    @Deprecated
+    public static RaiseException from(RubyException exception, IRubyObject backtrace) {
+        return new RaiseException(exception, backtrace);
     }
 
     private boolean requiresBacktrace(ThreadContext context) {

--- a/core/src/test/java/org/jruby/exceptions/TestRaiseException.java
+++ b/core/src/test/java/org/jruby/exceptions/TestRaiseException.java
@@ -9,6 +9,7 @@ import org.jruby.RubyClass;
 import org.jruby.RubyException;
 import org.jruby.RubyObject;
 import org.jruby.RubyRuntimeAdapter;
+import org.jruby.RubyString;
 import org.jruby.anno.JRubyMethod;
 import org.jruby.javasupport.JavaEmbedUtils;
 import org.jruby.runtime.Block;
@@ -361,6 +362,57 @@ public class TestRaiseException extends Base {
         public IRubyObject raise_from_nil(ThreadContext context) {
             throw RaiseException.from(context.runtime, context.runtime.getZeroDivisionError(), "raise_from_nil", context.nil);
         }
+    }
 
+    public void testNewRaiseExceptioin() {
+        Ruby ruby = Ruby.newInstance();
+
+        try {
+            throw ruby.newRaiseException(ruby.getException(), "blah");
+        } catch (Exception e) {
+            assertEquals("(Exception) blah", e.getMessage());
+        }
+
+        try {
+            throw ruby.newRaiseException(ruby.getException(), "blah");
+        } catch (Exception e) {
+            assertEquals("(Exception) blah", e.getMessage());
+        }
+
+        RubyArray backtrace = ruby.newEmptyArray();
+        try {
+            throw RaiseException.from(ruby, ruby.getException(), "blah", backtrace);
+        } catch (Exception e) {
+            assertEquals("(Exception) blah", e.getMessage());
+            assertEquals(backtrace, e.getException().backtrace());
+        }
+
+        RubyString message = ruby.newString("blah");
+        try {
+            throw RaiseException.from(ruby, ruby.getException(), message);
+        } catch (Exception e) {
+            assertEquals("(Exception) blah", e.getMessage());
+            assertEquals(message, e.getException().message(ruby.getCurrentContext()));
+        }
+
+        try {
+            throw RaiseException.from(ruby, "Exception", "blah");
+        } catch (Exception e) {
+            assertEquals("(Exception) blah", e.getMessage());
+        }
+
+        try {
+            throw RaiseException.from(ruby, "Exception", "blah", backtrace);
+        } catch (Exception e) {
+            assertEquals("(Exception) blah", e.getMessage());
+            assertEquals(backtrace, e.getException().backtrace());
+        }
+
+        try {
+            throw RaiseException.from(ruby, "Exception", message);
+        } catch (Exception e) {
+            assertEquals("(Exception) blah", e.getMessage());
+            assertEquals(message, e.getException().message(ruby.getCurrentContext()));
+        }
     }
 }

--- a/spec/java_integration/addons/throwable_spec.rb
+++ b/spec/java_integration/addons/throwable_spec.rb
@@ -128,6 +128,48 @@ describe "Rescuing a Java exception using Exception" do
     expect(i).to eq 2
   end
 
+  it "synchronizes causes" do
+    ex = Class.new(StandardError)
+
+    begin
+      e = ex.new
+
+      expect(e.cause).to be_nil
+
+      raise e
+    rescue
+      begin
+        e2 = ex.new
+
+        expect(e2.cause).to be_nil
+
+        t = JRuby.ref(e2)
+
+        expect(t.getCause).to be_nil
+
+        raise e2
+      rescue
+        expect(e2.cause).to equal(e)
+        expect(t.getCause).to equal(e)
+      end
+    end
+
+    begin
+      e = java.lang.NullPointerException.new
+
+      raise e
+    rescue java.lang.NullPointerException
+      begin
+        e2 = ex.new
+
+        raise e2
+      rescue
+        expect(e2.cause).to equal(e)
+        expect(JRuby.ref(e2).toThrowable.getCause).to equal(e)
+      end
+    end
+  end
+
   describe 'Ruby sub-class' do
 
     class RubyThrowable < java.lang.Exception

